### PR TITLE
Allow fetching unit by type, and build without type

### PR DESCRIPTION
**Pull request by BlueTheCube**
If your pull request is **not** translation or serverlist-related, read the list of requirements below and check each box:

An expansion of #9374 

- [X] I have read the [contribution guidelines](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md).
- [x] I have ensured that my code compiles, if applicable.
- [ ] I have ensured that any new features in this PR function correctly in-game, if applicable.

### CODE REVIEW

**Comment by BlueTheCube** [comment]
how did it fail

**Comment by BlueTheCube** [comment]
it has worked again

**Review by Anuken — CHANGES_REQUESTED** [review]

**Comment by BlueTheCube** [comment]
wanted

File: core/src/mindustry/logic/LExecutor.java [hunk 1]
```diff
@@ -1363,13 +1363,21 @@ public void run(LExecutor exec){
             TeamData data = t.data();
 
             switch(type){
-                case unit -> exec.setobj(result, i < 0 || i >= data.units.size ? null : data.units.get(i));
+                case unit -> {
+                    UnitType type = exec.obj(extra) instanceof UnitType u ? u : null;
+                    if(type == null){
+                        exec.setobj(result, i < 0 || i >= data.units.size ? null : data.units.get(i));
+                    }else{
+                        var units = data.unitCache(type);
```
Comment (Anuken) [inline core/src/mindustry/logic/LExecutor.java:1371]: `unitCache` can return null. This can crash.

Comment (BlueTheCube) [reply core/src/mindustry/logic/LExecutor.java:1371]: What about the `i < 0 || i >= data.units.size ? null : data.units.get(i)` below?

**Comment by BlueTheCube** [comment]
needs suggested code

File: core/src/mindustry/logic/LExecutor.java (hunk 1 above)
Comment (BlueTheCube) [reply core/src/mindustry/logic/LExecutor.java:1371]: I thought `unitCache` was `@Nullable`

Comment (BlueTheCube) [reply core/src/mindustry/logic/LExecutor.java:1371]: Fixed the `unitCount` nullpointerexception crash when there's no specific unit in extra

**Comment by BlueTheCube** [comment]
Closed due to commit bloat. Will reopen

**Comment by BlueTheCube** [comment]
No response, see #9703

**Pull request closed without merging**